### DISCUSSION TITLE
Recursive population for graph like entities after moving to v5

### DISCUSSION THREAD

**Discussion by tiaanvow** [mikro-orm/mikro-orm#4561] [Q&A]
**Describe the bug**
I am in the process of trying to update my org from mikro v4 to v5. I am running into an issue whereby population of graph like entities is behaving substantially differently to v4. Please bear with me as I try to explain what is happening.

**Entity Structure**
I have two entities, an `Ingredient` and a `MixingProcess`. The key fields that are relevant to this bug are:

For `Ingredient`
```
export class Ingredient {
// some other properties
@ManyToOne({ entity: () => MixingProcess, cascade: [Cascade.MERGE], nullable: true })
  fromMixingProcess?: MixingProcess;
```
For `MixingProcess`
```
export class MixingProcess {
// some other properties
// parents
  @ManyToMany({ entity: () => Ingredient, cascade: [Cascade.MERGE], nullable: true, owner: true })
  parents = new Collection<Ingredient>(this);

  // children
  @OneToMany({ entity: () => Ingredient, mappedBy: 'fromMixingProcess', cascade: [Cascade.MERGE], nullable: true })
  children = new Collection<Ingredient>(this);
```
These two entities combine to form a graph like structure:
```mermaid
graph LR

  classDef ingredient fill:#f9d5e5;
  classDef process fill:#14b1ab;

  subgraph Ingredients
    p1(Ingredient 1)
    p2(Ingredient 2)
    i1(Ingredient 4)
    i2(Ingredient 5)
    p3(Ingredient 3)
    c1(Ingredient 6)
    c2(Ingredient 7)
  end

  subgraph MixingProcesses
    mp1(MixingProcess 1) 
    mp2(MixingProcess 2)
  end

  p1 -->|parents| mp1
  p2 -->|parents| mp1
  mp1 -->|children| i1
  mp1 -->|children| i2
  i1 -->|fromMixingProcess| mp1
  i2 -->|fromMixingProcess| mp1
  i1 -->|parents| mp2
  i2 -->|parents| mp2
  p3 -->|parents| mp2
  mp2 -->|children| c1
  mp2 -->|children| c2
  c1 -->|fromMixingProcess| mp2
  c2 -->|fromMixingProcess| mp2

  class mp1,mp2 process
  class p1,p2,p3,i1,i2,c1,c2 ingredient

```

**Population issue**
The issue I am facing is that when I retrieve a list of `Ingredient`, usually we would populate `fromMixingProcess.parents.<property>` to extract some key property about the `Ingredient`s parents. In V4 this worked fine, and it would return only that field for the `fromMixingProcess.parents.<property>`. However now that I have updated to V5, the populate function retrieves `fromMixingProcess.parents` but then also `fromMixingProcess.parents.fromMixingProcess.parents.fromMixingProcess.parents` and so on. It continues all the way to the bottom of the tree. 

**To Reproduce**
```
let [ingredients, count] = await em.findAndCount(
    entities.Ingredient, { filters}, { limit, offset, orderBy})
await em.populate(ingredients, [`fromMixingProcess.parents.description`]);
```

**Expected behavior**
I would expect the populate not to continue fetching all of the sub properties for each of `populate` fields unless explicitly told to do so. e.g. if I wanted `fromMixingProcess.parents.fromMixingProcess.parents` then I would have to write the query as:
```
await em.populate(ingredients, [`fromMixingProcess.parents.fromMixingProcess.parents`]);
```

**Additional context**
* I am globally specifying the `populateWhere` property as: `populateWhere: PopulateHint.INFER`
* I am using an independent request context per request: 
```
router.use((_req, _res, next) => {
    RequestContext.create(router.context.postgres.em, next);
  });
```

**Versions**
| Dependency | Version |
| - | - |
| node | 18.16.1 |
| typescript | 4.9.5 |
| mikro-orm | 5.7.13 |
| your-driver | @mikro-orm/postgresql |

**Comment by B4nan** [comment]
Thanks for describing your problem in a detailed way, but I need to see a complete (but minimal) reproduction for such issues, I won't be able to reproduce anything from those tiny snippets. A code I can try that will clearly show what you think is wrong. It should as well prepare some data - like a test case. 

> await em.populate(ingredients, [`fromMixingProcess.parents.description`]);

Not sure what `description` is, but you can't populate anything else than a relation (or a lazy scalar property), this feels like you are confusing population and partial loading.

> cascade: [Cascade.MERGE]

This itself can have also some unwanted (but unrelated) side effects, first of all, cascade merging is no longer configurable (so this is a no-op, it's always enabled), second, by doing this, you disable cascading persist (which is almost never what people think/want).

> I would expect the populate not to continue fetching all of the sub properties for each of populate fields unless explicitly told to do so. e.g. if I wanted fromMixingProcess.parents.fromMixingProcess.parents then I would have to write the query as:

It works like that unless you have `populate: true` somewhere, there is no such thing as a recursive population outside of `populate: true`. People often confuse that because they see the entities forming a cycle, but that is based on their data - as there is always only one instance of a specific entity.

I am not entirely sure if you are talking about actual population issues (in other words, you see many unwanted queries in the logs - but I kinda doubt it), or if we are talking about serialization issues (so your API response looks different than you want to - those are already addressed in v6 via #4203).

Complete repro should help me understand better.
